### DISCUSSION TITLE
Type hinting for decorating custom dataclass with inheritance

### DISCUSSION THREAD

**Discussion by leverage-analytics** [microsoft/pyright#7726] [Q&A]
This is a cross-post from [this link](https://stackoverflow.com/questions/78349509/python-type-hinting-for-decorating-custom-dataclass-with-inheritance).

I'm curious where this should be working or not...

The problem is that the type checker is unable to recognize the methods from class A at the top of the resolution order from within instances of class Test, even when I explicitly give the return type.

I'm using Pyright with Pylance v2024.4.1 in VSCode.
```
from __future__ import annotations

import dataclasses
import typing as t
import types

_Self = t.TypeVar("_Self", bound="A")

class A:
  """Base class for all classes"""
  @classmethod
  def create(cls: type[_Self], *args, **kwargs) -> _Self:
    return cls(*args, **kwargs)

  def do_something(self) -> None:
    """a 'void' method"""
    pass


class B(A):
  """A secondary base class for the future dataclasses; NOTE: inheritance from A!"""


T = t.TypeVar("T", bound=B)
TCallable = t.Callable[[type], type[T]]


def decorator(**kwargs) -> TCallable:
  """A decorator for reducing boilerplate"""
  def _decorator(cls: type) -> type[B]:
    return types.new_class(
      cls.__name__, 
      (
        dataclasses.dataclass(**{"frozen": True, "kw_only": True, **kwargs})(cls), 
        B,
      ),
    )
  return _decorator


@decorator(repr=False)
class Test:
   """This is an implementation of the dataclass, subclassed from B, subclassed from A"""
   name: str
   
    def __repr__(self) -> str:
      return self.name
    
    def do_something_else(self) -> None:
      self.do_something() # <- not recognized by pyright

``

**Comment by erictraut** [comment]
There are a bunch of issues in this code.

1. This code is dynamically creating new classes. This won't work with static typing, as type checkers must be able to reason about classes statically.

2. The code is also defining a generic type alias `TCallable` that requires a single type argument, but when the type alias is used, no type argument is provided. It is therefore assumed to be `Any`.

3. The code is also syntactically incorrect. It's missing a closing parentheses in the call expression within the `_decorator` function.

4. The pre-decorated implementation of `Test` makes assumptions that the decorator will introduce a mix-in class (`B`).

So the short answer to your question is "no", this will not type check.

If you want your code to work with type checking, you should directly use the `dataclass` decorator and subclass from `B`, like this:

```python
@dataclass(frozen=True, kw_only=True, repr=False)
class Test(B):
    name: str
    
    def __repr__(self) -> str:
        return self.name
    
    def do_something_else(self) -> None:
        self.do_something()
```

**Reply by leverage-analytics** [reply]
Eric, thanks for the response. I'm sorry about the syntax errors; I was trying to providing a minimal example of what I was thinking.

I appreciate your feedback!
